### ISSUE TITLE
[BUG] midas_integrate_v2.binning.polygon.PolygonBinGeometry ignores detector tilt, distortion, panel shifts, and the residual-correction map

### ISSUE THREAD

**Issue by d-beniwal** [marinerhemant/MIDAS#69]
# `midas_integrate_v2.binning.polygon.PolygonBinGeometry` ignores detector tilt, distortion, panel shifts, and the residual-correction map

## Summary

`PolygonBinGeometry.from_spec()` (the `"polygon"` integration kernel, documented as
MIDAS's most accurate binning mode) computes each pixel's position using a bare
beam-centre-only projection (`Yc = -(Y - BC_y)`, `Zc = Z - BC_z`). It never calls
the package's own shared geometry function, `pixel_to_REta_from_spec`, that every
other kernel (`HardBinGeometry`, `SubpixelBinGeometry`, and
`corrections.integrated.integrate_with_corrections`) uses. As a result, the polygon
kernel silently drops:

- detector tilt (`tx`, `ty`, `tz`)
- the 15-coefficient distortion polynomial (`p0`..`p14` / v2 harmonic names)
- parallax correction
- per-panel rigid-body corrections (`δy`, `δz`, `δθ`, optionally `δLsd`/`δp₂`) for
  tiled multi-panel detectors (Pilatus/Eiger), read from `PanelShiftsFile`
- the empirical Stage-4 residual-correction spline (`ResidualCorrectionMap`)

There is no warning, no error, no log message — the geometry is simply wrong for
any detector that isn't perfectly untilted, undistorted, monolithic, and without a
residual map. This directly contradicts the module's own docstring claim
("`PolygonBinGeometry` — **exact**; differentiates MIDAS from every other
integrator") and its use elsewhere in the package as the accuracy reference
(`corrections/background.py`: *"use the polygon kernel (most accurate) for the
auto-fit"*).

**We found this while auditing whether `midas-gui`'s multi-panel calibration
feature correctly reaches the integration step — it does, for every kernel except
Polygon.**

## Environment

- Package: `midas-integrate-v2`
- Version tested: **0.6.0** (confirmed current/latest on PyPI as of 2026-08-28
  via `pip index versions midas-integrate-v2`)
- File: `midas_integrate_v2/binning/polygon.py`

## Root cause

`binning/polygon.py` line 51 imports both geometry functions:

```python
from ..forward import eval_pixel_REta, pixel_to_REta_from_spec
```

...but **neither is ever called anywhere in the file** (confirmed via
`grep -n "eval_pixel_REta\|pixel_to_REta_from_spec" binning/polygon.py` — the
only two hits are this import line). This strongly suggests the intent was to use
them and it was overlooked, rather than a deliberate simplification.

Instead, `PolygonBinGeometry.from_spec` (line 263) builds pixel corners via its own
private helper:

```python
# binning/polygon.py, line 223
def _pixel_corner_YZ_px(spec: IntegrationSpec) -> np.ndarray:
    """Same as ``_pixel_corner_YZ`` but in pixel units (so R = sqrt(Y²+Z²)
    is directly comparable to spec.RMin/RMax)."""
    NY, NZ = spec.NrPixelsY, spec.NrPixelsZ
    BC_y = float(spec.BC_y.detach())
    BC_z = float(spec.BC_z.detach())
    Yidx, Zidx = np.meshgrid(np.arange(NY), np.arange(NZ), indexing="xy")
    out = np.empty((NZ, NY, 4, 2), dtype=np.float64)
    for c, (dy, dz) in enumerate(_PIXEL_CORNERS):
        out[:, :, c, 0] = -(Yidx + dy - BC_y)
        out[:, :, c, 1] = (Zidx + dz - BC_z)
    return out
```

This only ever reads `spec.BC_y`/`spec.BC_z` — no `tx`/`ty`/`tz`, no distortion
coefficients, no `NPanelsY`/`PanelShiftsFile`, no `ResidualCorrectionMap`.

Compare with the shared function every other kernel uses
(`forward/pixels.py`, line 190):

```python
def pixel_to_REta_from_spec(Y_pix, Z_pix, spec):
    """...Applies the per-panel rigid shift and the per-panel Lsd / p2 offsets
    when the spec describes a multi-panel detector, matching the C
    DetectorMapper. Without this a 48-panel Pilatus integrates with its panel
    calibration silently discarded."""
    ...
    panels = _panel_inputs_from_spec(spec)          # reads PanelShiftsFile
    ...
    return _v2_pixel_to_REta(
        Y_pix, Z_pix,
        Lsd=spec.Lsd, BC_y=spec.BC_y, BC_z=spec.BC_z,
        tx=spec.tx, ty=spec.ty, tz=spec.tz,
        p_coeffs=_build_p_coeffs_from_spec(spec),
        parallax=spec.Parallax,
        ...
        residual_corr_map=_load_residual_map(spec),  # reads ResidualCorrectionMap
        **panel_kw,
    )
```

We confirmed (by reading every call site in the package — `cli.py`,
`streaming/integrate_stream.py`, `streaming/multi_detector.py`,
`pipelines/energy_sweep.py`, `corrections/background.py`) that nothing pre-warps
the image or spec before calling `PolygonBinGeometry.from_spec(spec)` to compensate
for this — the gap is inside `from_spec` itself, in every call path.

## Numeric proof

Comparing `PolygonBinGeometry`'s per-corner R (pixel radius) against
`pixel_to_REta_from_spec`'s R at the *exact same* sub-pixel position, on a
200×200 synthetic detector (`RBinSize=1.0 px`):

| Case                                            | max \|ΔR\| (px) | mean \|ΔR\| (px) |
|--------------------------------------------------|----------------:|------------------:|
| Baseline (no tilt/distortion/panel)              |           0.0000 |             0.0000 |
| Tilt only (`ty = 5°`)                             |           1.4577 |             0.3803 |
| Distortion only (`iso_R2 = 0.05`)                 |           0.2871 |             0.0627 |
| Multi-panel (two panels, `±2px/±3px` rigid shift) |           3.6055 |             2.2440 |

The baseline case matches exactly (0.0000 px), confirming the comparison
methodology is sound and isolating the discrepancy to exactly the three
corrections `PolygonBinGeometry` never applies. The multi-panel case's `max|ΔR|`
of 3.6055 px is `√(2² + 3²)` — i.e. `PolygonBinGeometry` is using the *unshifted*
panel position, off by exactly the injected panel shift.

With production `RBinSize` values often well under 1 px (sub-pixel R resolution is
explicitly called out as Polygon's target use case in its own docstring), a
1.5 px tilt-induced error or a 3.6 px panel-shift error is many bins wide — this
is not a rounding-level discrepancy.

<details>
<summary>Reproduction script (self-contained, run against midas-integrate-v2 0.6.0)</summary>

```python
import numpy as np
import torch
from midas_integrate_v2.spec import IntegrationSpec
from midas_integrate_v2.forward import pixel_to_REta_from_spec
from midas_integrate_v2.binning.polygon import _pixel_corner_YZ_px, _PIXEL_CORNERS

def make_spec(ty=0.0, tz=0.0, iso_R2=0.0, panel=False):
    s = IntegrationSpec()
    s.NrPixelsY = 200; s.NrPixelsZ = 200
    s.pxY = 200.0; s.pxZ = 200.0
    s.Lsd = torch.tensor(200000.0, dtype=torch.float64)
    s.BC_y = torch.tensor(100.0, dtype=torch.float64)
    s.BC_z = torch.tensor(100.0, dtype=torch.float64)
    s.tx = torch.tensor(0.0, dtype=torch.float64)
    s.ty = torch.tensor(ty, dtype=torch.float64)
    s.tz = torch.tensor(tz, dtype=torch.float64)
    s.Wavelength = torch.tensor(0.1729, dtype=torch.float64)
    s.Parallax = torch.tensor(0.0, dtype=torch.float64)
    s.iso_R2 = torch.tensor(iso_R2, dtype=torch.float64)
    s.RhoD = 141421.0
    s.RMin = 10.0; s.RMax = 137.0; s.RBinSize = 1.0
    s.EtaMin = -180.0; s.EtaMax = 180.0; s.EtaBinSize = 5.0
    if panel:
        s.NPanelsY = 1; s.NPanelsZ = 2
        s.PanelSizeY = 200; s.PanelSizeZ = 100
        s.PanelGapsY = []; s.PanelGapsZ = [0]
        s.PanelShiftsFile = "/tmp/_polygon_bug_shifts.txt"
        with open(s.PanelShiftsFile, "w") as f:
            f.write("    0  +2.000000  +3.000000  +0.000000e+00  +0.0000  +0.000000e+00\n")
            f.write("    1  -2.000000  -3.000000  +0.000000e+00  +0.0000  +0.000000e+00\n")
    return s

def compare(label, spec):
    corners = _pixel_corner_YZ_px(spec)
    poly_R_c0 = np.sqrt(corners[..., 0, 0]**2 + corners[..., 0, 1]**2)
    NY, NZ = spec.NrPixelsY, spec.NrPixelsZ
    dy, dz = _PIXEL_CORNERS[0]
    Yidx, Zidx = np.meshgrid(np.arange(NY), np.arange(NZ), indexing="xy")
    Y = torch.as_tensor(Yidx + dy, dtype=torch.float64)
    Z = torch.as_tensor(Zidx + dz, dtype=torch.float64)
    out = pixel_to_REta_from_spec(Y, Z, spec)
    true_R_c0 = out.R_px.detach().cpu().numpy()
    diff = np.abs(poly_R_c0 - true_R_c0)
    print(f"{label}: max|dR|={diff.max():.4f} px  mean|dR|={diff.mean():.4f}")

compare("Baseline",                          make_spec())
compare("Tilt only (ty=5 deg)",               make_spec(ty=5.0))
compare("Distortion only (iso_R2=0.05)",      make_spec(iso_R2=0.05))
compare("Multi-panel (+-2px/+-3px shift)",    make_spec(panel=True))
```

</details>

## Impact

- **Multi-panel detectors (Pilatus, Eiger)**: any tiled detector calibrated via
  `PanelLayout`/`PanelShiftsFile` has its per-panel rigid correction completely
  discarded when integrated with `mode="polygon"` — the module gaps/offsets are
  geometrically wrong at exactly the panel boundaries.
- **Any tilted detector**: `ty`/`tz` are core calibration outputs (virtually every
  real calibration refines them) — Polygon mode silently reverts to an untilted
  model.
- **Any distortion-corrected calibration**: the standard v2 harmonic distortion
  fit is dropped.
- **Any calibration with a kept Stage-4 residual spline**: also dropped.
- All of the above happen **with no warning, no exception, no log line** — a user
  has no way to discover this short of independently cross-checking against
  another kernel, which is how we found it.
- This is especially concerning because `PolygonBinGeometry` is the kernel
  documented and used internally as the *accuracy reference* — a user reasonably
  choosing it *because* it's marketed as most exact gets the geometrically least
  correct result of the three kernels whenever tilt, distortion, or panels are in
  play.

## Suggested fix

We see two tiers, because the three corrections are not equally tractable within
the existing exact-polygon-intersection algorithm:

**Panel shifts are a rigid transform and should be straightforward to fix
properly.** A per-panel `(δy, δz, δθ)` correction just translates/rotates the
pixel quad as a rigid body; both the pixel quad and the circular-arc bin
boundaries stay exactly representable in the (now panel-corrected) `(Yc, Zc)`
frame, so the existing Green's-theorem intersection math should carry over
directly once `_pixel_corner_YZ_px` applies each pixel's panel shift (via
`_panel_inputs_from_spec`/`panel_idx_for_points`, the same helpers
`pixel_to_REta_from_spec` already uses) before computing `Yc`/`Zc`.

**Tilt and distortion are harder**, because they are non-affine: lines of
constant `R` are no longer circles in raw pixel-index space once either is
nonzero, which breaks the "polygon vs. circular arc" shape assumption the exact
intersection algorithm (`pixel_bin_intersect`) relies on. Two options, in
increasing order of effort:

1. **Safety net now**: have `PolygonBinGeometry.from_spec` detect nonzero
   `tx`/`ty`/`tz`, nonzero distortion coefficients, a set `PanelShiftsFile`, or a
   set `ResidualCorrectionMap`, and raise/warn loudly (mirroring the existing
   `PanelShiftsMissingWarning` pattern in `forward/pixels.py`) — e.g. *"Polygon
   kernel does not apply tilt/distortion/panel/residual corrections; results will
   be geometrically incorrect for this spec. Use 'hard' or 'subpixel' instead, or
   proceed if you have verified these are negligible."* This at minimum turns a
   silent correctness bug into a visible, actionable one, and can ship
   immediately without touching the intersection math.
2. **Proper fix for tilt/distortion**: locally linearize `pixel_to_REta_from_spec`
   at each pixel centre (its Jacobian w.r.t. `Y_pix`/`Z_pix`) to get a per-pixel
   affine approximation of the true geometry, then run the existing exact
   intersection against that locally-linearized quad. This preserves "exact
   polygon area" as a *local* statement (exact for the linear approximation at
   that pixel) while extending correctness to the tilted/distorted case; the
   panel-shift fix above composes naturally with this since it's already affine.

We'd suggest doing the panel-shift fix and the safety-net warning as a fast
first pass (both are small, well-scoped changes), and treating the general
tilt/distortion case as a separate, larger follow-up.

## Files referenced

- `midas_integrate_v2/binning/polygon.py` — `_pixel_corner_YZ_px` (line 223),
  `PolygonBinGeometry.from_spec` (line 263)
- `midas_integrate_v2/forward/pixels.py` — `pixel_to_REta_from_spec` (line 190),
  `_panel_inputs_from_spec` (line 108), `_load_residual_map` (line 47)
- `midas_integrate_v2/corrections/integrated.py` — `integrate_with_corrections`
  (uses `pixel_to_REta_from_spec` directly, for comparison/contrast)

---
*Filed after auditing `midas-gui`'s multi-panel-detector calibration →
integration pipeline (2026-08-28); reproduced independently against a synthetic
`IntegrationSpec`, not tied to any midas-gui-specific code path.*

**Comment by marinerhemant** [comment]
Thanks for this — you were right, and the report was precise enough to reproduce directly. Fixed in **midas-integrate-v2 0.7.0**.

**Confirmed before fixing.** I reproduced your numbers independently on a 200×200 synthetic spec rather than taking them on trust, and they match to four decimals: baseline `max|ΔR| = 0.0000 px`, `ty = 5°` → **1.4577 px**, `iso_R2 = 0.05` → **0.2871 px**. Your panel figure of 3.6056 px is exactly √(2² + 3²), i.e. the unshifted panel position — so that one was the whole shift, not a partial error.

**Cause**, exactly as you diagnosed: `_pixel_corner_YZ_px` read only `spec.BC_y` / `spec.BC_z`. `eval_pixel_REta` and `pixel_to_REta_from_spec` were imported at the top of the module and never called, so tilt, the 15-coefficient distortion polynomial, parallax, per-panel rigid shifts and the Stage-4 residual map were all silently dropped — in the one kernel documented as "exact".

**Fix.** The polygon intersection already works in a frame where bins are exact circles (bin corners are built with `REta_to_YZ`), so only the pixel side was skipping the corrections. Pixel corners now map through the same shared path:

```
(Y_pix, Z_pix) --pixel_to_REta_from_spec--> (R_px, η) --REta_to_YZ--> (Y', Z')
```

That picks up tilt, distortion, parallax, panel shifts and the residual map in one change, because `pixel_to_REta_from_spec` already applies all of them. Panel index is computed once from pixel *centres* and reused for all four corners — otherwise a pixel straddling a panel boundary gets four different shifts and the quad tears.

**After the fix**: the polygon kernel matches the shared geometry to **2.8e-14 px**. On a real 48-panel Pilatus CeO2 frame the ring-radius-vs-azimuth spread went **0.513 → 0.345 px**.

**Two things worth flagging beyond the report:**

1. Chasing this turned up a second defect in the same area. `QUAD_ORDER = (0,1,3,2)` is a closed boundary only for Z-ordered corners, but both v1's constant and v2's own were boundary-ordered — so the shoelace walk traced two sides and two diagonals, a bowtie. Interior pixels were retaining a **mean of 0.85** of their area, worst case 0.50. Area conservation is now 1.4e-14 over 4000 pixels. Fixed in `midas-integrate 0.7.0`.

2. The docstring's "**exact** to fp64 precision — no approximation" is no longer literally true once corrections are applied, and now says so: it is exact polygon-vs-arc **in the corrected frame**, with a straight-chord pixel edge whose error is second order in pixel size — against the first-order 1.46 px error of ignoring the geometry entirely.

`area_weight=` is now selectable on `from_spec`: `"pixel"` (default) renormalises each pixel to unit weight, preserving agreement with the hard and subpixel kernels in the fine-bin limit; `"projected"` keeps the mapped quad area for solid-angle-like weighting.

Regression tests are in `tests/test_polygon_geometry_corrections.py`, including your three cases — they fail on the old code, so they discriminate rather than just pass.

**Comment by d-beniwal** [comment]
Thanks for the fix. I have tested it and also verified in the midas-gui development. Closing the issue.
